pkg/ingester/wal.go

```diff
@@ -0,0 +1,209 @@
+func (noopWAL) Stop()                {}
+
+type walWrapper struct {
+	cfg        WALConfig
+	wal        *wal.WAL
+	metrics    *ingesterMetrics
+	seriesIter SeriesIter
+
+	wait sync.WaitGroup
+	quit chan struct{}
+}
+
+// newWAL creates a WAL object. If the WAL is disabled, then the returned WAL is a no-op WAL.
+func newWAL(cfg WALConfig, registerer prometheus.Registerer, metrics *ingesterMetrics, seriesIter SeriesIter) (WAL, error) {
+	if !cfg.Enabled {
+		return noopWAL{}, nil
+	}
+
+	tsdbWAL, err := wal.NewSize(util.Logger, registerer, cfg.Dir, walSegmentSize, false)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &walWrapper{
+		cfg:        cfg,
+		quit:       make(chan struct{}),
+		wal:        tsdbWAL,
+		metrics:    metrics,
+		seriesIter: seriesIter,
+	}
+
+	w.wait.Add(1)
+	go w.run()
+	return w, nil
+}
+
+func (w *walWrapper) Log(record *WALRecord) error {
+	if record == nil || (len(record.Series) == 0 && len(record.RefEntries) == 0) {
+		return nil
+	}
+	select {
+	case <-w.quit:
+		return nil
+	default:
+		buf := recordPool.GetBytes()[:0]
+		defer func() {
+			recordPool.PutBytes(buf)
+		}()
+
+		// Always write series then entries.
+		if len(record.Series) > 0 {
+			buf = record.encodeSeries(buf)
+			if err := w.wal.Log(buf); err != nil {
+				return err
+			}
+			w.metrics.walRecordsLogged.Inc()
+			w.metrics.walLoggedBytesTotal.Add(float64(len(buf)))
+			buf = buf[:0]
+		}
+		if len(record.RefEntries) > 0 {
+			buf = record.encodeEntries(buf)
+			if err := w.wal.Log(buf); err != nil {
+				return err
+			}
+			w.metrics.walRecordsLogged.Inc()
+			w.metrics.walLoggedBytesTotal.Add(float64(len(buf)))
+		}
+		return nil
+	}
+}
+
+func (w *walWrapper) Stop() {
+	close(w.quit)
+	w.wait.Wait()
+	_ = w.wal.Close()
+	level.Info(util.Logger).Log("msg", "stopped", "component", "wal")
+}
+
+func (w *walWrapper) checkpointWriter() *WALCheckpointWriter {
+	return &WALCheckpointWriter{
+		metrics:    w.metrics,
+		segmentWAL: w.wal,
+	}
+}
+
+func (w *walWrapper) run() {
+	level.Info(util.Logger).Log("msg", "started", "component", "wal")
+	defer w.wait.Done()
+
+	checkpointer := NewCheckpointer(
+		w.cfg.CheckpointDuration,
+		w.seriesIter,
+		w.checkpointWriter(),
+		w.metrics,
+		w.quit,
+	)
+	checkpointer.Run()
+
+}
+
+type resettingPool struct {
+	rPool *sync.Pool // records
+	ePool *sync.Pool // entries
+	bPool *sync.Pool // bytes
+}
+
+func (p *resettingPool) GetRecord() *WALRecord {
+	rec := p.rPool.Get().(*WALRecord)
+	rec.Reset()
+	return rec
+}
+
+func (p *resettingPool) PutRecord(r *WALRecord) {
+	p.rPool.Put(r)
+}
+
+func (p *resettingPool) GetEntries() []logproto.Entry {
+	return p.ePool.Get().([]logproto.Entry)
+}
+
+func (p *resettingPool) PutEntries(es []logproto.Entry) {
+	p.ePool.Put(es[:0]) // nolint:staticcheck
+}
+
+func (p *resettingPool) GetBytes() []byte {
+	return p.bPool.Get().([]byte)
+}
+
+func (p *resettingPool) PutBytes(b []byte) {
+	p.bPool.Put(b[:0]) // nolint:staticcheck
```

Comment:
I think it might be better to use bytes.Buffer for those []byte slice, for the staticcheck lint reason.

You could benchmark it but basically the slice header ref is bigger than just a ref.

Reply:
I was considering using `*[]byte` for the alloc reason, but it seemed pretty minimal. Yes, it's bigger than a pointer, but it's very small due to the pointer-like semantics of slices.